pkg/virt-launcher/virtwrap/network/common.go

```diff
@@ -332,38 +332,30 @@ func (h *NetworkUtilsHandler) GetMacDetails(iface string) (net.HardwareAddr, err
 }
 
 // SetRandomMac changes the MAC address for a given interface to a randomly generated, preserving the vendor prefix
-func (h *NetworkUtilsHandler) SetRandomMac(iface string) (net.HardwareAddr, error) {
-	var mac net.HardwareAddr
-
-	currentMac, err := h.GetMacDetails(iface)
-	if err != nil {
-		return nil, err
-	}
-
-	changed := false
+func (h *NetworkUtilsHandler) SetRandomMac(iface netlink.Link) (net.HardwareAddr, error) {
+	ifaceName := iface.Attrs().Name
+	currentMac := iface.Attrs().HardwareAddr
 
 	for i := 0; i < randomMacGenerationAttempts; i++ {
-		changed, err = lmf.SpoofMacSameVendor(iface, false)
+		changed, err := lmf.SpoofMacSameVendor(ifaceName, false)
 		if err != nil {
 			log.Log.Reason(err).Errorf("failed to spoof MAC for an interface: %s", iface)
 			return nil, err
 		}
 
 		if changed {
-			mac, err = h.GetMacDetails(iface)
+			updatedIface, err := h.LinkByName(ifaceName)
 			if err != nil {
 				return nil, err
 			}
+			mac := updatedIface.Attrs().HardwareAddr
 			log.Log.Infof("updated MAC for %s interface: old: %s -> new: %s", iface, currentMac, mac)
-			break
+			return mac, nil
 		}
 	}
-	if !changed {
-		err := fmt.Errorf("failed to spoof MAC for an interface %s after %d attempts", iface, randomMacGenerationAttempts)
-		log.Log.Reason(err)
-		return nil, err
-	}
-	return currentMac, nil
+	err := fmt.Errorf("failed to spoof MAC for an interface %s after %d attempts", iface, randomMacGenerationAttempts)
+	log.Log.Reason(err)
+	return nil, err
 }
```

Comment:
The original algorithm has more clear responsabilities per block
1 - for loop until it changes
2 - handler return (error if not changed/mac if changed).

Reply:
for / if / break is better ?..

I can leave it as is, but I disagree. Anyway, the important thing is to get rid of the stale code.

Reply:
I'm wondering about plenty:
  - what is the purpose of the mac variable, defined outside the for loop ? It's only used inside it. At the very least, its scope is wrong.
  - why on earth does this return the old mac address ? The only caller discards that output parameter. Furthermore, any caller has the interface. They **already** know the old mac address.
  - the current code tries to change the MAC N times; if it succeeds, it sets a boolean variable that will be later on inspect to know how to return. Why do we need this variable altogether ? I can't for the life of me figure that out. The implementation of this method roughly translates in english to "check if the mac updater device managed to update the mac and return error if it didn't".

Reply:
Furthermore, how can it be OK to return an error from inside the loop, but not when I see the actual condition that I was looking for ? Why are errors an exception ? Should I, using your logic, mutating something else that I will check outside the loop to know how to return ?..

Reply:
@maiqueb I would refactor `SetRandomMac` in the following manner:
```code
// SetRandomMac changes the MAC address for a given interface to a randomly generated, preserving the vendor prefix
func (h *NetworkUtilsHandler) SetRandomMac(iface netlink.Link) (err error) {
	ifaceName := iface.Attrs().Name
	currentMac := iface.Attrs().HardwareAddr

	for i := 0; i < randomMacGenerationAttempts; i++ {
	    changed, err := lmf.SpoofMacSameVendor(ifaceName, false)
	    if changed || err != nil {
		if err != nil {
		    log.Log.Reason(err).Errorf("failed to spoof MAC for an interface: %s", ifaceName)
		}
		break
            }
	}

	if !changed && err == nil {
	    err = fmt.Errorf("failed to spoof MAC for an interface %s after %d attempts", ifaceName, randomMacGenerationAttempts)
	    log.Log.Reason(err)
	}

	return err
}
```
I think `SetRandomMac` doesn't need to exit with error if it fails to read interface data after MAC was updated.
I see that `lmf.SpoofMacSameVendor` gives us a honest answer in `changed`, since it checks using `MacChanged` whether the MAC was updated on the interface. I think it is redundant to produce a log of the old MAC and the new MAC.
if the caller wants to verbose about the change, it's up the the caller. `SetRandomMac` as the name states should only set a New Random MAC.

Reply:
The reason for the log's existence can be read in [0]. I feel uncomfortable removing it.

[0] - https://github.com/kubevirt/kubevirt/commit/ab53b18c17da91abcd420f134be9273b3332c471

Reply:
@maiqueb No problem, let's keep the log, but let's move to [here](https://github.com/kubevirt/kubevirt/blob/0ed56e62f4bac0aebc8564c99eba36eec5d342aa/pkg/virt-launcher/virtwrap/network/podinterface.go#L477) let's save the old mac before entering `SetRandomMac` and lets print the log if `SetRandomMac` returned `nil` error. I think it's worth to move the log for the sake of complexity reduction `SetRandomMac`.

Reply:
> @maiqueb I would refactor `SetRandomMac` in the following manner:
> 
> ```
> // SetRandomMac changes the MAC address for a given interface to a randomly generated, preserving the vendor prefix
> func (h *NetworkUtilsHandler) SetRandomMac(iface netlink.Link) (err error) {
> 	ifaceName := iface.Attrs().Name
> 
> 	for i := 0; i < randomMacGenerationAttempts; i++ {
> 	    changed, err := lmf.SpoofMacSameVendor(ifaceName, false)
> 	    if changed || err != nil {
> 		if err != nil {
> 		    log.Log.Reason(err).Errorf("failed to spoof MAC for an interface: %s", ifaceName)
> 		}
> 		break
>             }
> 	}
> 
> 	if !changed && err == nil {
> 	    err = fmt.Errorf("failed to spoof MAC for an interface %s after %d attempts", ifaceName, randomMacGenerationAttempts)
> 	    log.Log.Reason(err)
> 	}
> 
> 	return err
> }
> ```

The above does not work, since the variables `changed` and `err` are scoped inside the for. I'd have to set them as vars outside the loop, and mutate them inside the loop.

I personally feel that code should be as immutable as possible, since that makes it a lot more error prone - you stop having to understand where / why something changed (nothing changes; it is computed, evaluated, then thrown away once you leave their scope).

That is the overall reason I updated it as I did; nothing gets changed: variables are assigned, and a local decision is made.

> 
> I think `SetRandomMac` doesn't need to exit with error if it fails to read interface data after MAC was updated.

Well, I'm very unsure about this; this means that you've updated the MAC to XYZ, but got no clue as to what the new mac is. I'm aware we're only removing a loop in the network, but, knowing this mac is important for debugging.

> I see that `lmf.SpoofMacSameVendor` gives us a honest answer in `changed`, since it checks using `MacChanged` whether the MAC was updated on the interface. I think it is redundant to produce a log of the old MAC and the new MAC.
> if the caller wants to verbose about the change, it's up the the caller. `SetRandomMac` as the name states should only set a New Random MAC.
> 
> EDIT: you don't need `CurrentMAC` and I am not sure if we need to pass the whole link struct

I will follow your suggestion of moving the log outside the function, it does make it simpler.

Also, that enables us to simply pass the ifaceName rather than the whole link.